internal/controller/messaging/subscriber.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
+
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/application"
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/container"
+)
+
+type key string
+
+const (
+	correlationHeaderKey       key = common.CorrelationHeader
+	CommandRequestTopic            = "CommandRequestTopic"
+	CommandResponseTopicPrefix     = "CommandResponseTopicPrefix"
+)
+
+func SubscribeCommands(ctx context.Context, dic *di.Container) errors.EdgeX {
+	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+	messageBusInfo := container.ConfigurationFrom(dic.Get).MessageQueue
+	requestTopic := messageBusInfo.Topics[CommandRequestTopic]
+	responseTopicPrefix := messageBusInfo.Topics[CommandResponseTopicPrefix]
+
+	messages := make(chan types.MessageEnvelope)
+	messageErrors := make(chan error)
+	topics := []types.TopicChannel{
+		{
+			Topic:    requestTopic,
+			Messages: messages,
+		},
+	}
+
+	messageBus := bootstrapContainer.MessagingClientFrom(dic.Get)
+	err := messageBus.Subscribe(topics, messageErrors)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				lc.Infof("Exiting waiting for MessageBus '%s' topic messages", requestTopic)
+				return
+			case err = <-messageErrors:
+				lc.Error(err.Error())
+			case msgEnvelope := <-messages:
+				lc.Debugf("Command request received on message queue. Topic: %s, Correlation-id: %s ", requestTopic, msgEnvelope.CorrelationID)
+
+				// expected command request topic scheme: #/<service-name>/<device-name>/<command-name>/<method>
+				topicLevels := strings.Split(msgEnvelope.ReceivedTopic, "/")
+				length := len(topicLevels)
+				if length < 4 {
+					lc.Error("Failed to parse and construct response topic scheme, expected request topic scheme: '#/<service-name>/<device-name>/<command-name>/<method>'")
+					continue
+				}
+
+				// expected command response topic scheme: #/<service-name>/<device-name>/<command-name>/<method>
+				serviceName := topicLevels[length-4]
+				deviceName := topicLevels[length-3]
+				commandName := topicLevels[length-2]
+				method := topicLevels[length-1]
+				responseTopic := strings.Join([]string{responseTopicPrefix, serviceName, deviceName, commandName, method}, "/")
+
+				var responseEnvelope types.MessageEnvelope
+				if strings.EqualFold(method, "get") {
+					var rawQuery string
+					for k, v := range msgEnvelope.QueryParams {
+						if rawQuery != "" {
+							rawQuery += "&"
+						}
+						rawQuery += fmt.Sprintf("%s=%s", k, v)
+					}
+
+					ctx = context.WithValue(ctx, correlationHeaderKey, msgEnvelope.CorrelationID)
+					events, edgexErr := application.GetCommand(ctx, deviceName, commandName, rawQuery, dic)
+					if edgexErr != nil {
+						responseEnvelope = types.NewMessageEnvelopeWithError(msgEnvelope.RequestID, edgexErr.Error())
+						err = messageBus.Publish(responseEnvelope, responseTopic)
```

Comment:
Should also log the error that is being published as the response. Same in each of this blocks